### PR TITLE
[Enhancement] prune unnecessary CTEProperty calculate task

### PR DESCRIPTION
**Pull request by packy92**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #
After CTEs enterring into Memo, the Enforce phase adds CTE properties. With n CTEs, 2^n CTE properties are generated, and for each group, 2^n tasks are generated to calculate the best plan for each possible CTE property, causing plans to time out.

## Problem Summary(Required) ：

When the parent's required property are m CTEs and each group may only consume k CTEs, we can derive the required CTE properties of child is the intersection of m CTEs and k CTEs, this helps to avoid a large number of repeated calculations.

The modification approach is as follows:

When initializing the logical property of the group after entering the memo, collect information on the used CTEs. Used CTEs refer to the external CTE set that a group can reference. 
For anchors and producers, used CTEs do not include CTEs defined by themselves. 
For consumers, it includes CTEs defined by themselves and external CTEs that their children may reference. 
For other nodes, it is the set of all external CTEs referenced by their children.

In enforceAndcostTask, when deriving the required CTE properties for the child:
For the children of anchor nodes:
- The required CTE properties for producers = (parent's required CTEs) intersect (producer used CTEs).
- The required CTE properties for other children = ((parent's required CTEs) union (anchor CTE ID)) intersect (child used CTEs).

For the children of other nodes:
- The required CTE properties = (parent's required CTEs) intersect (child used CTEs).
    
Similarly, by using the used CTE information, we can improve the early pruning of invalid plans.

For a sql may timeout before, the trace is:
+---------------------------------------------------------------------------------------------+
| Explain String                                                                              |
+---------------------------------------------------------------------------------------------+
|  2448916ms|-- Total[1] 26823ms                                                              |
|  2448916ms|    -- Analyzer[1] 4ms                                                           |
|  2448921ms|    -- Transformer[1] 13ms                                                       |
|  2448935ms|    -- Optimizer[1] 26798ms                                                      |
|  2448935ms|        -- Optimizer.preprocessMvs[1] 0ms                                        |
|  2448936ms|        -- Optimizer.RuleBaseOptimize[1] 33ms                                    |
|  2448937ms|            -- Optimizer.RuleBaseOptimize.RewriteTreeTask[47] 30ms               |
|  2448970ms|        -- Optimizer.CostBaseOptimize[1] 26760ms                                 |
|  2448970ms|            -- Optimizer.CostBaseOptimize.OptimizeGroupTask[3348268] 647ms       |
|  2448970ms|            -- Optimizer.CostBaseOptimize.OptimizeExpressionTask[3532552] 4667ms |
|  2448970ms|            -- Optimizer.CostBaseOptimize.ExploreGroupTask[3349112] 172ms        |
|  2448970ms|            -- Optimizer.CostBaseOptimize.DeriveStatsTask[3532552] 169ms         |
|  2448971ms|            -- Optimizer.CostBaseOptimize.ApplyRuleTask[399] 2ms                 |
|  2448977ms|            -- Optimizer.CostBaseOptimize.EnforceAndCostTask[6926818] 12130ms    |
|  2475732ms|        -- Optimizer.PhysicalRewrite[1] 1ms                                      |
|  2475733ms|    -- ExecPlanBuild[1] 6ms                                                      |
+---------------------------------------------------------------------------------------------+

now, it's:
+---------------------------------------------------------------------------------------+
| Explain String                                                                        |
+---------------------------------------------------------------------------------------+
| 12239948ms|-- Total[1] 98ms                                                           |
| 12239948ms|    -- Analyzer[1] 3ms                                                     |
| 12239952ms|    -- Transformer[1] 9ms                                                  |
| 12239963ms|    -- Optimizer[1] 80ms                                                   |
| 12239963ms|        -- Optimizer.preprocessMvs[1] 0ms                                  |
| 12239963ms|        -- Optimizer.RuleBaseOptimize[1] 26ms                              |
| 12239964ms|            -- Optimizer.RuleBaseOptimize.RewriteTreeTask[47] 23ms         |
| 12239990ms|        -- Optimizer.CostBaseOptimize[1] 51ms                              |
| 12239990ms|            -- JoinReorderLeftDeep[22] 3ms                                 |
| 12239995ms|            -- Optimizer.CostBaseOptimize.OptimizeGroupTask[1126] 0ms      |
| 12239995ms|            -- Optimizer.CostBaseOptimize.OptimizeExpressionTask[1479] 2ms |
| 12239995ms|            -- Optimizer.CostBaseOptimize.ExploreGroupTask[1618] 0ms       |
| 12239995ms|            -- Optimizer.CostBaseOptimize.DeriveStatsTask[1479] 2ms        |
| 12239995ms|            -- Optimizer.CostBaseOptimize.ApplyRuleTask[464] 2ms           |
| 12240000ms|            -- Optimizer.CostBaseOptimize.EnforceAndCostTask[2528] 24ms    |
| 12240042ms|        -- Optimizer.PhysicalRewrite[1] 0ms                                |
| 12240043ms|    -- ExecPlanBuild[1] 4ms                                                |
+---------------------------------------------------------------------------------------+
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 76 / 76 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ExpressionContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/LogicalProperty.java|27|27|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 76 / 76 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ExpressionContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/LogicalProperty.java|27|27|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/LogicalProperty.java [hunk 1]
```diff
@@ -66,21 +80,56 @@ public OneTabletProperty oneTabletProperty() {
...
+
+        if (type == LOGICAL_CTE_ANCHOR) {
+            LogicalCTEAnchorOperator anchorOperator = (LogicalCTEAnchorOperator) expressionContext.getOp();
+            cteIds.addAll(expressionContext.getChildLogicalProperty(0).getUsedCTEs().getCteIds());
+            cteIds.addAll(expressionContext.getChildLogicalProperty(1).getUsedCTEs().getCteIds());
+            cteIds.remove(anchorOperator.getCteId());
+        } else if (type == LOGICAL_CTE_PRODUCE) {
+            LogicalCTEProduceOperator produceOperator = (LogicalCTEProduceOperator) expressionContext.getOp();
+            cteIds.addAll(expressionContext.getChildLogicalProperty(0).getUsedCTEs().getCteIds());
+            cteIds.remove(produceOperator.getCteId());
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/LogicalProperty.java:119]: the children of producer never support producer's cteid

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/LogicalProperty.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/LogicalProperty.java:119]: Yes, the next remove step is unnecessary.

**Comment by Seaven** [comment]
I thought of another way:
Like the previous shuffle property, pass the CTE/NO-CTE property from the bottom up and check it at the anchor node.

Do you think this reduces property derivation?

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 74 / 74 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ExpressionContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/LogicalProperty.java|25|25|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 74 / 74 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ExpressionContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/LogicalProperty.java|25|25|100.00%|[]|

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java [hunk 2]
```diff
@@ -163,7 +165,7 @@ private Pair<QueryDumpInfo, String> getPlanFragment(String dumpJsonStr, SessionV
         if (sessionVariable != null) {
             queryDumpInfo.setSessionVariable(sessionVariable);
         }
-        queryDumpInfo.getSessionVariable().setOptimizerExecuteTimeout(30000000);
+        queryDumpInfo.getSessionVariable().setOptimizerExecuteTimeout(3000);
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/ReplayFromDumpTest.java:168]: Keep it large to make FE UT stable.

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 74 / 74 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|38|38|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ExpressionContext.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/LogicalProperty.java|25|25|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**